### PR TITLE
Interactive window

### PR DESCRIPTION
**Pull request by paulvanbrenk**
* Make sure we don't hit a null arg exception when doing telemetry
* add info command to interactive window to help debugging issues

### CODE REVIEW

File: Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    [Export(typeof(IReplCommand))]
+    internal class InfoReplCommand : IReplCommand
+    {
+        public Task<ExecutionResult> Execute(IReplWindow window, string arguments)
+        {
+            if (window.Evaluator is NodejsReplEvaluator nodeEval)
+            {
+                try
+                {
+                    var nodeExePath = nodeEval.NodeExePath;
```
Comment (armanio123) [inline Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs]: Prefer conditional operator
var nodeExePath = string.IsNullOrEmpty(nodeEval.NodeExePath) ? nodeEval.GetNodeExePath() : nodeEval.NodeExePath;

File: Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                    if (string.IsNullOrEmpty(nodeExePath))
```
Comment (armanio123) [inline Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs]: IsNullOrWhiteSpace preferred or necessary?

File: Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs (hunk 1 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs]: no

File: Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs (hunk 2 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs]: no, we probably want to fail in that case, and give the user a clear error message

File: Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                    {
+                        nodeExePath = nodeEval.GetNodeExePath();
+                    }
+                    var nodeVersion = FileVersionInfo.GetVersionInfo(nodeExePath);
+
+                    window.WriteLine(string.Format(CultureInfo.CurrentUICulture, Resources.ReplNodeInfo, nodeExePath));
+                    window.WriteLine(string.Format(CultureInfo.CurrentUICulture, Resources.ReplNodeVersion, nodeVersion.ProductVersion));
+                }
+                catch(Exception e)
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs]: Is this boilerplate or do we actually expect an exception?  If the main failure mode is a bad path, it seems like detecting it and printing a message to that effect would be more helpful than a stack trace.

File: Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using System.Globalization;
+using System.Threading.Tasks;
+
+namespace Microsoft.NodejsTools.Repl
+{
+    [Export(typeof(IReplCommand))]
+    internal class InfoReplCommand : IReplCommand
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs]: Nit: `sealed`

File: Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs [hunk 5]
```diff
@@ -39,6 +39,8 @@ public NodejsReplEvaluator(INodejsReplSite site)
             this._site = site;
         }
 
+        public string NodeExePath { get; private set; }
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs]: Shouldn't this compute and then cache `GetNodeExePath`?  Why would that be the caller's responsibility?

File: Nodejs/Product/TestAdapter/TestExecutor.cs [hunk 6]
```diff
@@ -195,7 +195,7 @@ private void LogTelemetry(int testCount, Version nodeVersion, bool isDebugging)
             //todo: when we have support for the Node 8 debugger log which version of the debugger people are actually using
 
             var defaultSession = TelemetryService.DefaultSession;
```
Comment (amcasey) [inline Nodejs/Product/TestAdapter/TestExecutor.cs]: This could probably be inlined. :)

File: Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs (hunk 3 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/Repl/InfoReplCommand.cs]: We already check when setting the path in a bunch of places. This is a backstop to not crash the interactive window when trying to get diagnostics.

File: Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs [hunk 7]
```diff
@@ -39,6 +39,20 @@ public NodejsReplEvaluator(INodejsReplSite site)
             this._site = site;
         }
 
+        private string nodeExePath;
+
+        public string NodeExePath
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(this.nodeExePath))
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs]: Should this be a straight null check?  Wouldn't empty string indicate that it had already run (and failed)?

File: Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs [hunk 8]
```diff
@@ -39,6 +39,20 @@ public NodejsReplEvaluator(INodejsReplSite site)
...
+                {
+                    this.nodeExePath = this.GetNodeExePath();
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs]: Is any thread safety required here?

File: Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs [hunk 9]
```diff
@@ -143,6 +152,30 @@ private void EnsureConnected()
...
         }
 
+        /// <summary>
+        /// Checks if Node.js Exe is installed correctly.
+        /// Writes an error message if it's not.
+        /// </summary>
+        /// <returns></returns>
+        public bool EnsureNodeInstalled()
+        {
+            if (string.IsNullOrWhiteSpace(this.NodeExePath))
```
Comment (amcasey) [inline Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs]: Null is unexpected, correct?

File: Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs (hunk 7 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs]: Yes, but when a new project is loaded we want to try again, maybe they configured the Node location there correct.

File: Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs (hunk 8 above)
Comment (paulvanbrenk) [reply Nodejs/Product/Nodejs/Repl/NodejsReplEvaluator.cs]: no, initialization is all on the UI thread.

**Pull request merged**